### PR TITLE
p2p/adapters: Use RPC with SimNodes

### PR DESCRIPTION
**Pull request by lmars**
This changes `SimNode` to wrap a `node.Service` and provide an RPC client for the service which brings it inline with the API for interacting with the exec and docker adapters.

This will allow us to expose the RPC client to simulation tests so that they can make arbitrary assertions on the nodes using RPC regardless of the adapter in use.

I have made `p2p.Server` an interface which `SimNode` now implements, and added `SimAdminAPI` which is analogous to `node.PrivateAdminAPI`.

### CODE REVIEW

File: cmd/swarm/main.go [hunk 1]
```diff
@@ -440,7 +440,7 @@ func getPassPhrase(prompt string, i int, passwords []string) string {
 	return password
 }
 
-func injectBootnodes(srv *p2p.Server, nodes []string) {
+func injectBootnodes(srv p2p.Server, nodes []string) {
```
Comment (zelig) [inline cmd/swarm/main.go:443]: so you changed p2p.Server to an interface which is implemented by p2p.server (rlpx server) and also adapters.SimNode? 
genious

File: p2p/adapters/inproc.go [hunk 2]
```diff
@@ -112,69 +180,99 @@ func (self *SimNode) setPeer(id *NodeId, rw *p2p.MsgPipeRW) *Peer {
...
+
+func (self *SimNode) PeersInfo() (info []*p2p.PeerInfo) {
 	return nil
 }
 
-func (self *SimNode) RunProtocol(id *NodeId, rw, rrw p2p.MsgReadWriter, peer *Peer) error {
-	if self.Run == nil {
+func (self *SimNode) RunProtocol(node *SimNode, rw, rrw p2p.MsgReadWriter, peer *Peer) {
+	id := node.Id
+	protocol := self.service.Protocols()[0]
```
Comment (zelig) [inline p2p/adapters/inproc.go:238]: so why not allow multiple protocols to start then?

File: p2p/simulations/network.go [hunk 3]
```diff
@@ -37,6 +37,7 @@ import (
 	"github.com/ethereum/go-ethereum/log"
 	"github.com/ethereum/go-ethereum/p2p/adapters"
 	"github.com/ethereum/go-ethereum/p2p/discover"
+	"github.com/ethereum/go-ethereum/rpc"
 )
 
 type NetworkConfig struct {
```
Comment (zelig) [inline p2p/simulations/network.go:43]: this is beautiful now

File: p2p/testing/protocoltester.go [hunk 4]
```diff
@@ -5,79 +5,84 @@ import (
...
 
 }
+
+// testNode wraps a protocol run function and implements the node.Service
+// interface
+type testNode struct {
+	run func(*p2p.Peer, p2p.MsgReadWriter) error
+}
+
+func (t *testNode) Protocols() []p2p.Protocol {
```
Comment (zelig) [inline p2p/testing/protocoltester.go:74]: we could allow multiple protocols here for generality

File: swarm/network/simulations/overlay.go [hunk 5]
```diff
@@ -39,26 +48,29 @@ func af() <-chan time.Time {
 }
 
 // Start() starts up the hive
-// makes SimNode implement *NodeAdapter
-func (self *SimNode) Start() error {
-	return self.hive.Start(self.connect, af)
+// makes SimNode implement node.Service
+func (self *SimNode) Start(server p2p.Server) error {
```
Comment (zelig) [inline swarm/network/simulations/overlay.go:52]: try to eliminate the duplication of this bit?

File: p2p/adapters/inproc.go [hunk 6]
```diff
@@ -112,69 +180,99 @@ func (self *SimNode) setPeer(id *NodeId, rw *p2p.MsgPipeRW) *Peer {
...
+	if protocol.Run == nil {
 		log.Trace(fmt.Sprintf("no protocol starting on peer %v (connection with %v)", self.Id, id))
-		return nil
+		return
 	}
 	log.Trace(fmt.Sprintf("protocol starting on peer %v (connection with %v)", self.Id, id))
 	p := p2p.NewPeer(id.NodeID, id.Label(), []p2p.Cap{})
 	go func() {
 		self.network.DidConnect(self.Id, id)
```
Comment (zelig) [inline p2p/adapters/inproc.go:246]: so we are still calling DIdConnect here. It is ok
For other types we need to implement a subscription via IPC/WS to signal the connection events.
In which case inproc can just behave the same, analogous to connect (via admin_addPeer)

File: swarm/network/simulations/discovery/discovery_test.go [hunk 7]
```diff
@@ -26,7 +26,9 @@ const serviceName = "discovery"
 func init() {
 	// register the discovery service which will run as a devp2p
 	// protocol when using the exec adapter
-	adapters.RegisterService(serviceName, discoveryService)
+	adapters.RegisterService(serviceName, func(id *adapters.NodeId) p2pnode.Service {
```
Comment (zelig) [inline swarm/network/simulations/discovery/discovery_test.go:28]: is this surely ok to have this global? is it always in global testing scope?

File: p2p/adapters/inproc.go (hunk 2 above)
Comment (lmars) [reply p2p/adapters/inproc.go:238]: I was just trying to make a small change.

If there are multiple protocols running we need to have a mechanism to stop them if one of them exits before the others.

This is doable, but not necessary at the moment given we aren't testing anything which runs multiple protocols. We should just modify this when the time arises that we need to support multiple protocols.

File: p2p/adapters/inproc.go (hunk 6 above)
Comment (lmars) [reply p2p/adapters/inproc.go:246]: Yes I plan to tackle events separately now that we have RPC for all adapters.

File: swarm/network/simulations/overlay.go (hunk 5 above)
Comment (lmars) [reply swarm/network/simulations/overlay.go:52]: I removed the duplication in 7a718fe.

File: swarm/network/simulations/discovery/discovery_test.go (hunk 7 above)
Comment (lmars) [reply swarm/network/simulations/discovery/discovery_test.go:28]: It is fine for now yes, the code will panic on `init` if two packages try to register the same name, in which case we can just rename things.

File: p2p/adapters/inproc.go (hunk 2 above)
Comment (zelig) [reply p2p/adapters/inproc.go:238]: point taken

**Pull request merged**